python/lsst/meas/base/_id_generator.py

```diff
@@ -0,0 +1,548 @@
+            Configuration for an ID generator.
+        fixed : `DataCoordinate`
+            Data ID identifying the dimensions that are considered fixed by the
+            `IdGenerator` that produced the IDs: usually just ``instrument`` or
+            ``skymap``, depending on the configuration. For most configurations
+            this will need to be a fully-expanded data ID.
+
+        Returns
+        -------
+        unpacker
+            Callable that takes a single `int` argument (an ID generated by an
+            identically-configured `IdGenerator`) and returns a tuple of:
+
+            - release_id: the integer that identifies a data release or
+              similar (`int`);
+            - data_id : the data ID used to initialize the original ID
+              generator (`DataCoordinate`);
+            - counter : the counter part of the original ID (`int`).
+
+        Notes
+        -----
+        This method cannot be used on IDs generated without a data ID.
+        """
+        packer = config._make_dimension_packer(fixed)
+        return cls.unpacker_from_dimension_packer(packer, config.n_releases)
+
+    @classmethod
+    def unpacker_from_dimension_packer(
+        cls,
+        dimension_packer: DimensionPacker,
+        n_releases: int = DEFAULT_N_RELEASES,
+    ) -> Callable[[int], tuple[int, DataCoordinate, int]]:
+        """Return a callable that unpacks the IDs generated by this class,
+        from a `lsst.daf.butler.DimensionPacker` instance.
+
+        Parameters
+        ----------
+        dimension_packer : `lsst.daf.butler.DimensionPacker`
+            Dimension packer used to construct the original
+            `DimensionPackerIdGenerator`.
+        n_releases : `int`, optional
+            Number of (contiguous, starting from zero) ``release_id`` values to
+            reserve space for. One (not zero) is used to reserve no space.
+
+        Returns
+        -------
+        unpacker
+            Callable that takes a single `int` argument (an ID generated by an
+            identically-constructed `DimensionPackerIdGenerator`) and returns a
+            tuple of:
+
+            - release_id: the integer that identifies a data release or
+              similar (`int`);
+            - data_id : the data ID used to initialize the original ID
+              generator (`DataCoordinate`);
+            - counter : the counter part of the original ID (`int`).
+
+        Notes
+        -----
+        This method cannot be used on IDs generated with no data ID.
+        """
+        bits = _IdGeneratorBits(dimension_packer, n_releases)
+
+        def unpack(record_id: int) -> tuple[int, DataCoordinate, int]:
+            rest, counter = divmod(record_id, bits.n_counters)
+            rest, packed_data_id = divmod(rest, bits.n_data_ids)
+            rest, release_id = divmod(rest, bits.n_data_ids)
+            if rest:
+                raise ValueError(f"Unexpected overall factor {rest} in record_id {record_id}.")
```

Comment:
Maybe include the other unpacked things for debugging? e.g. `... in record_id {record_id}, after unpacking counter={counter}, packed_data_id={packed_data_id}, release_id={release_id}`

Reply:
I think most of the ways to get to this error involve those being garbage, because you're trying to unpack something generated some other way, but sure, it can't hurt.